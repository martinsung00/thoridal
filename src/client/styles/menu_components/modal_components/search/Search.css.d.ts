export const mainLogo: string;
export const searchBar: string;
export const searchType: string;
export const searchBy: string;
export const searchButton: string;
export const searchButtonGray: string;
export const divider: string;
export const noHistory: string;
