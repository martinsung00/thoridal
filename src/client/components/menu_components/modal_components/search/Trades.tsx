import React from "react";
import styles from "../../../../styles/menu_components/modal_components/search/Trades.css";
import { Trade } from "../../../../types/index";

interface Props {
  trades: Array<Trade> | [];
  page: number;
  pages: number;
  handleDelete: Function;
}

export const Trades: React.FunctionComponent<Props> = (props: Props) => {
  return (
    <div className={styles.root}>
      {!props.trades.length && (
        <div>
          <div className={styles.errorState}>
            Could Not Find Any Relevant Trades
          </div>
          <br />
          <div className={styles.errorText}>Try Searching Again!</div>
        </div>
      )}
      {props.trades.length !== 0 && (
        <div style={{ marginTop: "-4em" }}>
          {(props.trades as Trade[]).map((trade: Trade, index: number) => {
            return (
              <div key={index} className={styles.trade}>
                <div className={styles.content}>
                  <div className={styles.info}>
                    {" "}
                    <div>
                      Trade ID:
                      <div className={styles.item}>{trade.id}</div>
                    </div>
                    <br />
                    <div>
                      Reference Number:
                      <div className={styles.item}>
                        {trade.reference_number}
                      </div>
                    </div>
                    <br />
                    <div>
                      Ticker:
                      <div className={styles.item}>{trade.ticker}</div>
                    </div>
                    <br />
                    <div>
                      Company Name:
                      <div className={styles.item}>{trade.company_name}</div>
                    </div>
                    <br />
                    <br />
                  </div>
                  <div className={styles.note}>
                    <div className={styles.comment}>{trade.note}</div>
                    <br />
                    <div className={styles.additionalInfo}>
                      <div className={styles.details}>
                        {trade.trade_action === "bought" ? "Bought" : "Sold"}
                      </div>
                      <div className={styles.details}>
                        {trade.trade_type === "short" ? "Short" : "Long"}
                      </div>
                      <div className={styles.details}>
                        {trade.trade_status ? "Fulfilled" : "Not Fulfilled"}
                      </div>
                      <div className={styles.details}>
                        {`${trade.created_at}`.slice(0, 10)}
                      </div>
                      <br />
                      <div className={styles.cost}>
                        <div className={styles.details}>
                          Total Cost: ${trade.total_cost}
                        </div>
                        <div className={styles.details}>
                          Unit Price: ${trade.unit_price}
                        </div>
                        <div className={styles.details}>
                          Quantity: {trade.quantity}
                        </div>
                      </div>
                      <div className={styles.controls}>
                        <div
                          className={styles.controlUpdate}
                          id="update-button"
                        >
                          {"\u21ea"}
                        </div>
                        <div
                          className={styles.controlDelete}
                          id="delete-button"
                          onClick={function (
                            event: React.MouseEvent<HTMLDivElement, MouseEvent>
                          ) {
                            const id = trade.id;
                            props.handleDelete(event, index, id);
                          }}
                        >
                          {"\u2297"}
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            );
          })}
          {props.trades.length > 10 && (
            <div className={styles.next}>Next {"\u21e2"}</div>
          )}
          {props.trades.length < 10 && props.trades.length > 0 && (
            <div className={styles.nextGray}>Next {"\u21e2"}</div>
          )}
          {props.page > 1 && <div className={styles.prev}>{"\u21e0"} Prev</div>}
          {props.page === 1 && (
            <div className={styles.prevGray}>{"\u21e0"} Prev</div>
          )}
        </div>
      )}
    </div>
  );
};
