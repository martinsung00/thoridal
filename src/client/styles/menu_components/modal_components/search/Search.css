.mainLogo {
  position: relative;
  height: 35px;
  width: 35px;
  margin-left: 47.5%;
}

.searchBar {
  position: relative;
  width: 20em;
  height: 1em;
  font-size: 22px;
  font-family: Helvetica, Arial, sans-serif;
  top: 3em;
  left: 26%;
  border: 3px solid gray;
  border-radius: 3px;
}

.searchType {
  position: relative;
  margin-right: 36em;
  margin-top: 4.8em;
  float: right;
  background-color: rgba(255, 255, 255, 0.952);
  padding-left: 3%;
  padding-right: -1%;
  border-left: 0.7px solid darkgray;
  width: 2.5em;
  font-family: Helvetica, Arial, sans-serif;
  font-size: 15px;
  outline: none;
}

.searchBy {
  border: none;
  outline: none;
  font-size: 14.5px;
  font-weight: bold;
  color: rgb(116, 116, 116);
  background-color: white;
  cursor: pointer;
}

.searchButton {
  position: relative;
  text-align: center;
  padding-bottom: 0.25%;
  left: 19.5%;
  float: right;
  width: 4em;
  height: 1em;
  font-size: 22px;
  font-family: Helvetica, Arial, sans-serif;
  top: 3em;
  color: white;
  background-color: black;
  border: 3px solid black;
  border-radius: 3px;
  cursor: pointer;
}

.searchButtonGray {
  position: relative;
  text-align: center;
  padding-bottom: 0.25%;
  left: 19.5%;
  float: right;
  width: 4em;
  height: 1em;
  font-size: 22px;
  font-family: Helvetica, Arial, sans-serif;
  top: 3em;
  color: rgb(0, 0, 0);
  background-color: rgb(99, 99, 99);
  border: 3px solid rgb(100, 100, 100);
  border-radius: 3px;
  cursor: pointer;
}

.searchButton:hover {
  position: relative;
  text-align: center;
  padding-bottom: 0.25%;
  left: 19.5%;
  float: right;
  width: 4em;
  height: 1em;
  font-size: 22px;
  font-family: Helvetica, Arial, sans-serif;
  top: 3em;
  color: rgb(0, 0, 0);
  background-color: rgb(180, 180, 180);
  border: 3px solid rgb(180, 180, 180);
  border-radius: 3px;
  cursor: pointer;
}

.divider {
  position: relative;
  height: 2em;
  top: 9em;
  left: 3.7em;
  width: 90%;
  position: relative;
  text-align: center;
  border-top: 0.7px solid black;
  border-bottom: 0.7px solid black;
}

.noHistory {
  position: relative;
  top: 5em;
  text-align: center;
  font-size: 35px;
  font-family: Helvetica, Arial, sans-serif;
  margin-top: 2em;
}
