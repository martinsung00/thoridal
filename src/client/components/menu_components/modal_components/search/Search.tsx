import React from "react";
import axios, { AxiosResponse } from "axios";
import styles from "../../../../styles/menu_components/modal_components/search/Search.css";
import { Trade } from "../../../../types/index";
import { Trades } from "./Trades";

interface Props {
  toggleShow: Function;
  show: boolean;
}

interface State {
  current_page: "search" | "confirm" | "failed";
  errored: boolean;
  searchTerm: string;
  searchType: string;
  trades: Array<Trade> | [];
  page: number;
  pages: number;
}

export default class ModalContent extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      current_page: "search",
      errored: false,
      searchTerm: "",
      searchType: "id",
      trades: [],
      page: 1,
      pages: 0,
    };
  }

  componentDidMount(): void {
    const component = this;

    axios
      .get("/trade/all/find")
      .then(function (response) {
        const data: Array<Trade> = response.data.rows;

        component.setState(
          {
            trades: data.reverse(),
            errored: false,
          },
          function () {
            const pages = Math.ceil(component.state.trades.length / 10);

            component.setState({
              pages: pages,
            });
          }
        );
      })
      .catch(function (error: Error) {
        console.error(error);
        component._handleError();
      });
  }

  public _handleSelect(event: React.ChangeEvent<HTMLSelectElement>): void {
    event.preventDefault();

    this.setState({
      searchType: event.currentTarget.value,
    });
  }

  public _handleChange(event: React.ChangeEvent<HTMLInputElement>): void {
    event.preventDefault();

    this.setState({
      searchTerm: event.target.value,
    });
  }

  public _handleDelete(
    event: React.MouseEvent<HTMLDivElement, MouseEvent>,
    index: number,
    id: string
  ): void {
    event.preventDefault();

    const component = this;

    axios
      .delete(`/trade/id/${id}/delete`)
      .then(function () {
        component.state.trades.splice(index, 1);
        const data = component.state.trades;
        component.setState({
          trades: data,
        });
      })
      .catch(function (error: Error) {
        console.error(error);
        component._handleError();
      });
  }

  public _handleError(): void {
    this.setState({
      errored: true,
    });
  }

  public _handleSearch(event: React.FormEvent): void {
    event.preventDefault();

    const component = this;
    const type = this.state.searchType;
    const term = this.state.searchTerm;

    axios
      .get(`/trade/${type}/${term}/find`)
      .then(function (response: AxiosResponse) {
        const data: Array<Trade> = response.data.rows;
        component.setState({
          trades: data,
          errored: false,
        });
      })
      .catch(function (error) {
        console.error(error);
        component._handleError();
      });
  }

  render() {
    return (
      <div>
        <div>
          <input
            placeholder="Search Here!"
            required
            className={styles.searchBar}
            id="search-by"
            value={this.state.searchTerm}
            onChange={this._handleChange.bind(this)}
          ></input>
          <div className={styles.searchType}>
            <select
              className={styles.searchBy}
              value={this.state.searchType}
              defaultValue="id"
              onChange={this._handleSelect.bind(this)}
            >
              <option value="id">ID</option>
              <option value="ticker">Ticker</option>
              <option value="company">Company</option>
              <option value="date">Date</option>
            </select>
          </div>
          {this.state.searchTerm.length > 0 && (
            <div
              className={styles.searchButton}
              onClick={this._handleSearch.bind(this)}
            >
              Search
            </div>
          )}
          {this.state.searchTerm.length <= 0 && (
            <div className={styles.searchButtonGray}>Search</div>
          )}
        </div>
        <div className={styles.divider} />
        {this.state.trades.length > 0 && (
          <Trades
            page={this.state.page}
            pages={this.state.pages}
            trades={this.state.trades}
            handleDelete={this._handleDelete.bind(this)}
          />
        )}
        {this.state.trades.length === 0 && (
          <div className={styles.noHistory}>No Previous Trade History</div>
        )}
      </div>
    );
  }
}
