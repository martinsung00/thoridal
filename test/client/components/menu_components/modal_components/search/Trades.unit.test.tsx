import React from "react";
import renderer from "react-test-renderer";
import { Trades } from "../../../../../../src/client/components/menu_components/modal_components/search/Trades";
import { Trade } from "../../../../../../src/client/types/index";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

interface Props {
  trades: Array<Trade> | [];
  page: number;
  pages: number;
  handleDelete: Function;
}

describe("Modal Component Trades Tests", function () {
  const props: Props = {
    trades: [
      {
        id: "TEST",
        ticker: "TEST",
        company_name: "TEST",
        reference_number: "TEST",
        unit_price: 0.0,
        quantity: 0,
        total_cost: 0.0,
        trade_type: "short",
        note: "TEST",
        created_at: new Date("1995-12-17T03:24:00"),
        trade_status: false,
        trade_action: "bought",
      },
    ],
    page: 1,
    pages: 1,
    handleDelete: jest.fn(),
  };

  const propsV2: Props = {
    trades: [
      {
        id: "TEST",
        ticker: "TEST",
        company_name: "TEST",
        reference_number: "TEST",
        unit_price: 0.0,
        quantity: 0,
        total_cost: 0.0,
        trade_type: "long",
        note: "TEST",
        created_at: new Date(),
        trade_status: true,
        trade_action: "sold",
      },
    ],
    page: 1,
    pages: 1,
    handleDelete: jest.fn(),
  };

  it("should render the elements as is", function () {
    const container:
      | renderer.ReactTestRendererJSON
      | renderer.ReactTestRendererJSON[]
      | null = renderer.create(<Trades {...props} />).toJSON();

    expect(container).toMatchSnapshot();
  });

  it("should render the trade's elements with 'long' and 'Fulfilled'", function () {
    const container:
      | renderer.ReactTestRendererJSON
      | renderer.ReactTestRendererJSON[]
      | null = renderer.create(<Trades {...propsV2} />).toJSON();

    expect(container).toMatchSnapshot();
  });

  it("should render the grayed out version of the next button if the current trades are empty", function () {
    const alternateProps: Props = {
      trades: [],
      page: 1,
      pages: 1,
      handleDelete: jest.fn(),
    };
    const wrapper = shallow(<Trades {...alternateProps} />);

    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
    jest.resetAllMocks();
  });

  it("should render the the prev button if the current page is > 1", function () {
    const alternateProps: Props = {
      trades: [
        {
          id: "TEST",
          ticker: "TEST",
          company_name: "TEST",
          reference_number: "TEST",
          unit_price: 0.0,
          quantity: 0,
          total_cost: 0.0,
          trade_type: "short",
          note: "TEST",
          created_at: new Date(),
          trade_status: false,
          trade_action: "bought",
        },
      ],
      page: 52,
      pages: 52,
      handleDelete: jest.fn(),
    };
    const wrapper = shallow(<Trades {...alternateProps} />);

    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
    jest.resetAllMocks();
  });

  it("should render the the next button if the trade length is > 1", function () {
    const alternateProps: Props = {
      trades: [
        {
          id: "TEST",
          ticker: "TEST",
          company_name: "TEST",
          reference_number: "TEST",
          unit_price: 0.0,
          quantity: 0,
          total_cost: 0.0,
          trade_type: "short",
          note: "TEST",
          created_at: new Date(),
          trade_status: false,
          trade_action: "bought",
        },
        {
          id: "TEST",
          ticker: "TEST",
          company_name: "TEST",
          reference_number: "TEST",
          unit_price: 0.0,
          quantity: 0,
          total_cost: 0.0,
          trade_type: "short",
          note: "TEST",
          created_at: new Date(),
          trade_status: false,
          trade_action: "sold",
        },
        {
          id: "TEST",
          ticker: "TEST",
          company_name: "TEST",
          reference_number: "TEST",
          unit_price: 0.0,
          quantity: 0,
          total_cost: 0.0,
          trade_type: "short",
          note: "TEST",
          created_at: new Date(),
          trade_status: false,
          trade_action: "bought",
        },
        {
          id: "TEST",
          ticker: "TEST",
          company_name: "TEST",
          reference_number: "TEST",
          unit_price: 0.0,
          quantity: 0,
          total_cost: 0.0,
          trade_type: "short",
          note: "TEST",
          created_at: new Date(),
          trade_status: false,
          trade_action: "sold",
        },
        {
          id: "TEST",
          ticker: "TEST",
          company_name: "TEST",
          reference_number: "TEST",
          unit_price: 0.0,
          quantity: 0,
          total_cost: 0.0,
          trade_type: "short",
          note: "TEST",
          created_at: new Date(),
          trade_status: false,
          trade_action: "sold",
        },
        {
          id: "TEST",
          ticker: "TEST",
          company_name: "TEST",
          reference_number: "TEST",
          unit_price: 0.0,
          quantity: 0,
          total_cost: 0.0,
          trade_type: "short",
          note: "TEST",
          created_at: new Date(),
          trade_status: false,
          trade_action: "sold",
        },
        {
          id: "TEST",
          ticker: "TEST",
          company_name: "TEST",
          reference_number: "TEST",
          unit_price: 0.0,
          quantity: 0,
          total_cost: 0.0,
          trade_type: "short",
          note: "TEST",
          created_at: new Date(),
          trade_status: false,
          trade_action: "sold",
        },
        {
          id: "TEST",
          ticker: "TEST",
          company_name: "TEST",
          reference_number: "TEST",
          unit_price: 0.0,
          quantity: 0,
          total_cost: 0.0,
          trade_type: "short",
          note: "TEST",
          created_at: new Date(),
          trade_status: false,
          trade_action: "sold",
        },
        {
          id: "TEST",
          ticker: "TEST",
          company_name: "TEST",
          reference_number: "TEST",
          unit_price: 0.0,
          quantity: 0,
          total_cost: 0.0,
          trade_type: "short",
          note: "TEST",
          created_at: new Date(),
          trade_status: false,
          trade_action: "sold",
        },
        {
          id: "TEST",
          ticker: "TEST",
          company_name: "TEST",
          reference_number: "TEST",
          unit_price: 0.0,
          quantity: 0,
          total_cost: 0.0,
          trade_type: "short",
          note: "TEST",
          created_at: new Date(),
          trade_status: false,
          trade_action: "sold",
        },
        {
          id: "TEST",
          ticker: "TEST",
          company_name: "TEST",
          reference_number: "TEST",
          unit_price: 0.0,
          quantity: 0,
          total_cost: 0.0,
          trade_type: "short",
          note: "TEST",
          created_at: new Date(),
          trade_status: false,
          trade_action: "sold",
        },
      ],
      page: 1,
      pages: 2,
      handleDelete: jest.fn(),
    };

    const wrapper = shallow(<Trades {...alternateProps} />);

    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
    jest.resetAllMocks();
  });

  it("should render the the prev button if the current page is > 1", function () {
    const alternateProps: Props = {
      trades: [],
      page: 2,
      pages: 2,
      handleDelete: jest.fn(),
    };
    const wrapper = shallow(<Trades {...alternateProps} />);

    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
    jest.resetAllMocks();
  });

  describe("Delete Button", function () {
    it("should call the handle delete method", async function () {
      const flushPromises = () => new Promise(setImmediate);
      const wrapper = shallow(<Trades {...props} />);
      const spy = jest.spyOn<any, any>(props, "handleDelete");

      wrapper.find("#delete-button").simulate("click");
      await flushPromises();
      wrapper.update();

      expect(spy).toHaveBeenCalled();

      wrapper.unmount();
      jest.resetAllMocks();
    });
  });
});
