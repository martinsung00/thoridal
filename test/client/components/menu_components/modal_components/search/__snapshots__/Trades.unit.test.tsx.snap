// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component Trades Tests should render the elements as is 1`] = `
<div>
  <div
    style={
      Object {
        "marginTop": "-4em",
      }
    }
  >
    <div>
      <div>
        <div>
           
          <div>
            Trade ID:
            <div>
              TEST
            </div>
          </div>
          <br />
          <div>
            Reference Number:
            <div>
              TEST
            </div>
          </div>
          <br />
          <div>
            Ticker:
            <div>
              TEST
            </div>
          </div>
          <br />
          <div>
            Company Name:
            <div>
              TEST
            </div>
          </div>
          <br />
          <br />
        </div>
        <div>
          <div>
            TEST
          </div>
          <br />
          <div>
            <div>
              Bought
            </div>
            <div>
              Short
            </div>
            <div>
              Not Fulfilled
            </div>
            <div>
              Sun Dec 17
            </div>
            <br />
            <div>
              <div>
                Total Cost: $
                0
              </div>
              <div>
                Unit Price: $
                0
              </div>
              <div>
                Quantity: 
                0
              </div>
            </div>
            <div>
              <div
                id="update-button"
              >
                ⇪
              </div>
              <div
                id="delete-button"
                onClick={[Function]}
              >
                ⊗
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      Next 
      ⇢
    </div>
    <div>
      ⇠
       Prev
    </div>
  </div>
</div>
`;

exports[`Modal Component Trades Tests should render the grayed out version of the next button if the current trades are empty 1`] = `ShallowWrapper {}`;

exports[`Modal Component Trades Tests should render the the next button if the trade length is > 1 1`] = `ShallowWrapper {}`;

exports[`Modal Component Trades Tests should render the the prev button if the current page is > 1 1`] = `ShallowWrapper {}`;

exports[`Modal Component Trades Tests should render the the prev button if the current page is > 1 2`] = `ShallowWrapper {}`;

exports[`Modal Component Trades Tests should render the trade's elements with 'long' and 'Fulfilled' 1`] = `
<div>
  <div
    style={
      Object {
        "marginTop": "-4em",
      }
    }
  >
    <div>
      <div>
        <div>
           
          <div>
            Trade ID:
            <div>
              TEST
            </div>
          </div>
          <br />
          <div>
            Reference Number:
            <div>
              TEST
            </div>
          </div>
          <br />
          <div>
            Ticker:
            <div>
              TEST
            </div>
          </div>
          <br />
          <div>
            Company Name:
            <div>
              TEST
            </div>
          </div>
          <br />
          <br />
        </div>
        <div>
          <div>
            TEST
          </div>
          <br />
          <div>
            <div>
              Sold
            </div>
            <div>
              Long
            </div>
            <div>
              Fulfilled
            </div>
            <div>
              Wed Feb 10
            </div>
            <br />
            <div>
              <div>
                Total Cost: $
                0
              </div>
              <div>
                Unit Price: $
                0
              </div>
              <div>
                Quantity: 
                0
              </div>
            </div>
            <div>
              <div
                id="update-button"
              >
                ⇪
              </div>
              <div
                id="delete-button"
                onClick={[Function]}
              >
                ⊗
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      Next 
      ⇢
    </div>
    <div>
      ⇠
       Prev
    </div>
  </div>
</div>
`;
